[Yeo Jun Wei] iP

JunWei3112 opened the pull request:
## Duke
> One of the secrets of getting more done is to make a 'to do list' every day, keep it visible and use it as a guide to action as   you go through the day - Jean De La Fontaine

Duke helps to keep track of the things you need to do, so that you don't have to! It's
* Text-based
* Easy to learn
* ~~Fast~~ SUPER FAST to use!!
* Most importantly, ***FREE!!!!***

To use Duke, you will have to do the following steps:
1. Download the file from [here](https://github.com/JunWei3112/ip/tree/master).
2. Double-click it.
3. Add your tasks.
4. Let Duke manage your tasks! :smile:

Features:
- [x] Managing tasks, deadlines and events
- [x] Loads and saves tasks in your hard disk
- [ ] Find deadlines/events occurring on a specific date (coming soon!)
- [ ] Reminders (coming soon!)

If you are a Java programmer, you can use it to practice Java too. (wow!)
Here's the main method: 
```java
public class Main  {
    public static void main(String[] args) {
        Application.launch(Duke.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
Greets user, echos commands and exits when user types "bye"

Commit:
Stores tasks entered by user, and displays them when requested

Commit:
Able to mark tasks as done

Commit:
Add support for tracking ToDos, Events and Deadlines

Commit:
Automate text UI testing, and fix errors found during automated testing

Commit:
Handle user errors, and remove version tag from all files

Commit:
Add support for deleting tasks, and handle more user errors

Commit:
Update automated tests to account for error handling and task deletion

Commit:
Add support for loading and saving tasks in hard disk

Commit:
Add support for understanding dates and times

Commit:
Merge branch 'branch-Level-7'

Commit:
Merge branch 'branch-Level-8'

Commit:
Update input and expected outcome for automated testing

Commit:
Refactor the code to extract out closely related code as classes

Commit:
Divide classes into suitable java packages

Commit:
Move Duke.java under src/main/java/duke

Commit:
Add JUnit tests

Commit:
Package the app as a JAR file

Commit:
Implement the isExit() method in *Command classes

Commit:
Add JavaDoc comments

Commit:
Tweak code to comply with Java coding standard

Commit:
Allow users to find tasks using a keyword

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-Level-9'

Commit:
Remove redundant method in Ui.java

Commit:
Merge branch 'add-gradle-support'

Commit:
Set up Gradle

Commit:
Merge branch 'branch-A-Gradle'

Commit:
Implement CheckStyle

Commit:
Merge branch 'branch-A-CheckStyle'

Commit:
Add a GUI to Duke

Commit:
Merge branch 'branch-Level-10'

lwj1711 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,114 @@
+                }
+
+                // Parsing the user input based on the task type
+                if (taskType.equals("deadline")) {
+                    if (!taskDescription.contains(" /by ")) {
+                        throw new TimeNotSpecifiedException(" /by ");
+                    }
+                    String[] descriptionArr = taskDescription.split(" /by ");
+                    task = descriptionArr[0];
+                    String deadline = descriptionArr[1];
+                    if (deadline.trim().isEmpty()) {
+                        throw new TimeNotSpecifiedException(" /by ");
+                    }
+                    date = LocalDate.parse(deadline);
+                    return new AddCommand(taskType, task, date);
+                } else if (taskType.equals("event")) {
+                    if (!taskDescription.contains(" /at ")) {
+                        throw new TimeNotSpecifiedException(" /at ");
+                    }
+                    String[] descriptionArr = taskDescription.split(" /at ");
+                    task = descriptionArr[0];
+                    String timeFrame = descriptionArr[1];
+                    if (timeFrame.trim().isEmpty()) {
+                        throw new TimeNotSpecifiedException(" /at ");
+                    }
+                    date = LocalDate.parse(timeFrame);
+                    return new AddCommand(taskType, task, date);
+                } else {
+                    task = taskDescription;
+                    return new AddCommand(taskType, task);
+                }
+            } else {
+                throw new UnrecognisedCommandException();
+            }
```
The logic in this block looks rather convoluted. Perhaps the use of switch statements can help?

JunWei3112 replied:
Good point, I will make changes to this block of code! Thanks for the catch, and also for the suggestion! 😄

lwj1711 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,43 @@
+ */
+public class Deadline extends Task {
+
+    protected LocalDate by;
```
I thought this could be a little ambiguous. Perhaps a name such as 'timeBy' might be better.

JunWei3112 replied:
Good catch, I will change this variable name as per your suggestion! Thanks! 👍🏽

lwj1711 reviewed (commented):
Looks great generally, only extremely minor nits on some parts. 🦄

Zhou-Jiahao-1998 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,78 @@
+/**
+ * This AddCommand class represents a command to add a task to the task list.
+ */
+public class AddCommand extends Command {
```
AddCommand sounds like an action. Will CommandAdder sound more like a class? Same goes for other subclasses of Command.

zhoukerrr commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,50 @@
+
+        // Displays a message indicating the task has been successfully deleted from the list
+        String message = "Got it! I've removed this task:\n" + "  " + removedTask + "\n";
+        if (tasks.getNumberOfTasks() <= 1) {
+            message += "Now you have " + tasks.getNumberOfTasks() + " task in the list.";
+        } else {
+            message += "Now you have " + tasks.getNumberOfTasks() + " tasks in the list.";
+        }
```
Just a nitpick, "Now you have " can probably be abstracted out and placed together in line 38.

JunWei3112 replied:
That's a good catch, I will make the following changes! Thanks! 😁

zhoukerrr reviewed (commented):
Overall very well done. Classes are abstracted and categorised clearly. Exceptions are well handled as well.

zhoukerrr commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,43 @@
+ */
+public class Event extends Task {
+
+    protected LocalDate at;
```
Maybe can be more specific? Similar point as Deadline.

JunWei3112 replied:
Good point, I will change this variable name too! Thanks! 😄

Commit:
Fix coding standard violations

Commit:
Implement assertions

Assumptions about the program state are not verified.

Missing out on possible bugs in the code due to incorrect invariants
can result in unsafe execution, where further code works with wrong
values and can potentially cause damage.

Let's define assertions to document and check important assumptions
that should hold at various points in the code.

Using assertions is preferable over exceptions in this situation since
we are checking for mistakes made by the programmer, not the
user or environment.

The build.gradle file has also been modified to fix the problem of
Gradle not being able to launch the GUI application.

Commit:
Improve code quality

Some areas of the code are either not readable, follow unsafe
practices, or have redundant and poor code comments.

These can cause the code to have poor quality, which makes it harder
to maintain and extend the code in the future.

Let's follow the provided 'Code Quality' topics to make
improvements to our existing code.

We closely follow the provided guidelines, as opposed to our
intuition, since intuition about code quality differ among
all programmers.

JunWei3112 committed:
Merge pull request #1 from JunWei3112/branch-A-Assertions

Implement assertions

Commit:
Merge branch 'master' into branch-A-CodeQuality

JunWei3112 committed:
Merge pull request #2 from JunWei3112/branch-A-CodeQuality

Improve code quality

Commit:
Add support for tagging tasks

Commit:
Merge branch 'branch-C-Tagging'

Commit:
Fix coding standard violations

Commit:
Update JUnit tests

Commit:
Fix GUI bug where list gets cut off

Commit:
Adjust vertical padding of the dialog box

Commit:
Merge branch 'branch-A-BetterGui'

Commit:
Include a representative screenshot of Duke

Commit:
Add user guide

Commit:
Improve on code quality

Commit:
Add a title to the Window

Commit:
Change data file path to use relative paths

Commit:
Add 'Quick Start' section to user guide

JunWei3112 committed:
Update to JavaFX 16 to fix font issue

JunWei3112 committed:
Update 'Quick Start' section of User Guide

JunWei3112 committed:
Set theme jekyll-theme-cayman

JunWei3112 committed:
Update _config.yml
